### ISSUE TITLE
Using wxWebViewChromium with CMake `RelWithDebugInfo` build type doesn't work under Linux

### ISSUE THREAD

**Issue by vadz** [wxWidgets/wxWidgets#24322]
This is nasty because in this situation, i.e. when you build an application using wx with `-DCMAKE_BUILD_TYPE=RelWithDebugInfo`, you just get run-time crash because `libcef_dll_wrapper` is compiled _without_ `-DNDEBUG`, as this is only done for `Release` build type by libcef makefiles, but `wxHAVE_CEF_DEBUG` is not defined in this case, so wx assumes that it was used.

I have no idea whether we can force the use of `Release` for a subdirectory when the build type is `RelWithDebugInfo`, but if we can't, we need to either define `NDEBUG` manually ourselves in this case, which is ugly, or, on the contrary, define `wxHAVE_CEF_DEBUG`, which is not ideal for the production build.

@MaartenBent Your ideas about this would be very welcome, of course, TIA!

**Comment by MaartenBent** [comment]
> I have no idea whether we can force the use of Release for a subdirectory when the build type is RelWithDebugInfo

Not sure if I understand the problem, is it just using the release libs and you want the debug libs? I think you can do this by adding `*_RELWITHDEBINFO` with the correct library to:
https://github.com/wxWidgets/wxWidgets/blob/63a40637ebe9f8944721fc448e46f979f4ebcdb4/build/cmake/lib/webview_chromium/CMakeLists.txt#L93-L99

For completeness maybe `MINSIZEREL` too.
And I don't know if it will be a problem on macOS, but you can add it to the `APPLE` case too.

**Comment by vadz** [comment]
Sorry, let me try to explain it again:

My own application uses wxWidgets from a submodule via `add_subdirectory()`, i.e. like this

```cmake
	set(wxUSE_WEBVIEW_CHROMIUM ON)
	add_subdirectory(3rdparty/wx EXCLUDE_FROM_ALL)

        ...

	target_link_libraries(MyApp
		PRIVATE
			wx::webview
	)
```

If I configure it using `-DCMAKE_BUILD_TYPE=RelWithDebugInfo`, it crashes during run-time due to the mismatch between CEF headers (compiled with `NDEBUG` defined, as wx itself is built as in "Release") and the CEF debug library being used in this case (compiled without `NDEBUG` defined).

Please let me know if you still don't see the problem?

**Comment by vadz** [comment]
BTW, on a maybe related note, I have to do

```cmake
target_link_libraries(MyApp_cef_helper PRIVATE $<TARGET_FILE:libcef_dll_wrapper>)
```

instead of just linking with `libcef_dll_wrapper` as CMake doesn't find the library otherwise and I don't really understand why -- is this somehow related to the location properties by chance?

**Comment by MaartenBent** [comment]
I'll try to reproduce it, and see what's going on with both problems.

**Comment by vadz** [comment]
Sorry, I have to leave now so I can't make an actually minimal repro myself, but at least for the first problem I think it's clear enough what is going: I've looked at the commands used to build CEF in this case and they don't include `-DNDEBUG`, while we do define it when including CEF headers from wx in this build. So we're using different ABIs with the predictable results.

I still hate that CEF prefers to silently crash rather than detect this :-(

**Comment by MaartenBent** [comment]
Small update after I struggled with linking errors for two hours. Apparently this
https://github.com/wxWidgets/wxWidgets/blob/7150e6ecb385f6b76c7e2d81ff030305db51b344/build/cmake/lib/webview_chromium/CMakeLists.txt#L83
overwrites all the compile options set by `libcef_dll_wrapper` (including `-fPIC`), instead of appending to it. I think we should've used `target_compile_options`.

**Comment by vadz** [comment]
Oops, sorry for breaking this. I am using autoconf and not CMake for testing wx and linking wx statically in my real application (which does use CMake), so I hadn't noticed it, sorry.

**Comment by MaartenBent** [comment]
Still can't reproduce it. I made the webview sample a stand-alone app, and added wxWidgets with `add_subdirectory`.
Then I built everything (including downloading and building libcef_dll_wrapper) with `-DCMAKE_BUILD_TYPE=RelWithDebInfo`. All builds fine, except the custom command to copy the cef data:
https://github.com/wxWidgets/wxWidgets/blob/7150e6ecb385f6b76c7e2d81ff030305db51b344/build/cmake/samples/webview.cmake#L20-L26
Because it uses `$<CONFIG>` which evaluates to  `RelWithDebInfo`, and cef doesn't have this folder. Only `Debug` and `Release`. When I make it use `Release` it builds fine. And the sample runs fine with the chromium backend too (after forcing x11 with `GDK_BACKEND=x11`. The weird thing is, when I use the `Debug` cef library it also runs fine...

It would really help if you could provide a minimal reproducible example including build commands.

I think I understand the `NDEBUG` problem, but just can't seem to reproduce it now. A fix could be:
* use `get_target_property(libcef_dll_wrapper ...)` to get the `COMPILE_DEFINITIONS` and `COMPILE_FLAGS` and `LINK_FLAGS_RELEASE`.
* For the first two use a regex or something to extract the release flags/options in `$<$<CONFIG:Release>:thereleaseonlyflags>`
* Set those for the `RelWithDebInfo` build: `$<$<CONFIG:RelWithDebInfo>:thereleaseonlyflags`>, and set the `LINK_FLAGS_RELEASE` to `LINK_FLAGS_RELWITHDEBINFO`.

I don't know why you need to link to `libcef_dll_wrapper` at all. This is part of the webview dependencies.

Also you say you used `RelWithDebugInfo`. The correct config name is `RelWithDebInfo`. Was this just a typo when creating this issue? Otherwise that could explain the lack of `NDEBUG` in your build. Because the compile flags for  `RelWithDebInfo` should contain this.
<img width="383" alt="Screenshot 2024-02-18 114512" src="https://github.com/wxWidgets/wxWidgets/assets/8088070/0170131f-6e42-48b7-8f36-7d399b3b5edd">

**Comment by vadz** [comment]
> Still can't reproduce it.

I'm awfully sorry, you were right and my problems were due to using the wrong build type, i.e., as you say

> Also you say you used `RelWithDebugInfo`. The correct config name is `RelWithDebInfo`. Was this just a typo when creating this issue? Otherwise that could explain the lack of `NDEBUG` in your build.

It was just due to this and if I use `-DCMAKE_BUILD_TYPE=RelWithDebInfo` I do see `-DNDEBUG` in `ninja -t commands libcef_dll_wrapper` output and, unsurprisingly, there are no crashes any more. Sorry again!

> I made the webview sample a stand-alone app, and added wxWidgets with `add_subdirectory`. Then I built everything (including downloading and building libcef_dll_wrapper) with `-DCMAKE_BUILD_TYPE=RelWithDebInfo`. All builds fine, except the custom command to copy the cef data:
> 
> https://github.com/wxWidgets/wxWidgets/blob/7150e6ecb385f6b76c7e2d81ff030305db51b344/build/cmake/samples/webview.cmake#L20-L26
> 
> Because it uses `$<CONFIG>` which evaluates to `RelWithDebInfo`, and cef doesn't have this folder. Only `Debug` and `Release`. When I make it use `Release` it builds fine. And the sample runs fine with the chromium backend too (after forcing x11 with `GDK_BACKEND=x11`. The weird thing is, when I use the `Debug` cef library it also runs fine...

I do understand this, actually: CEF debug and release libraries seem to be ABI-compatible. It's only `libcef_dll_wrapper` which isn't.

So we should probably just always use hardcoded `Release` when copying files.

> It would really help if you could provide a minimal reproducible example including build commands.

I don't think it's needed any more, sorry again for wasting your time.

> I don't know why you need to link to `libcef_dll_wrapper` at all. This is part of the webview dependencies.

I only link the helper process, which doesn't use wx at all, with it explicitly.

---

To try to salvage at least something from all this, should we perhaps check if the build type is one of the known strings? I think I should even be able to implement this myself, if you think it's a good idea...

Sorry once again!

**Comment by MaartenBent** [comment]
> To try to salvage at least something from all this, should we perhaps check if the build type is one of the known strings? I think I should even be able to implement this myself, if you think it's a good idea...

Yes, maybe a warning when an unknown build type is used. Not sure if to show it always or only when `wxUSE_WEBVIEW_CHROMIUM` is enabled, where it can do actual harm.

And we should definitely fix the issue in https://github.com/wxWidgets/wxWidgets/issues/24322#issuecomment-1949420656. Can you replace that with:
```cmake
target_compile_options(libcef_dll_wrapper PRIVATE "-Wno-extra")
```

> So we should probably just always use hardcoded Release when copying files.

I'm fine with this.
Then it would also work when users download the minimal cef distribution that only contains the release lib.
We could actually do this in the CMake download script too... No need for the standard distribution if we only use the release lib.

-----

There are still improvements to be made. I can try to extract the `Release` options from `libcef_dll_wrapper` and apply them to `RelWithDebInfo` and `MinSizeRel`. So they at least have the build settings as intended by the cef devs.

I also tried this with msvc and there the problem is worse. `libcef_dll_wrapper` explicitly does
```cmake
set(CMAKE_CONFIGURATION_TYPES Debug Release)
```
so when building this in `RelWithDebInfo`, Visual Studio doesn't find a matching configuration and fails to build.
I managed to work around this using `variable_watch`, and when `libcef_dll_wrapper` sets the value, I reset it to the original.
